### PR TITLE
[feature] Implement an Enumerables module

### PR DESCRIPTION
**Pull request by Mindo-Joseph**
### What does this PR do?
This PR creates an enumerable module and implements the following methods:
1. my_each
2. my_each_with_index
3. my_select
4. my_all?
5. my_none?
6. my_count
7. my_any?
8. my_map
9. my_inject

#### How should this be manually tested?
- Click on the repl.it button in the README.md file and run the given code examples at the bottom of the enumerables.rb file

### CODE REVIEW

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
+module Enumerable
+  def my_each
+    length.times { |i| yield(self[i]) }
+  end
```
Comment (suretrust) [inline enumerables.rb]: * [x] should have working `my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_each.class != [1,2,3,4,5].each.class
```

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+  def my_each_with_index
+    (0..length - 1).each do |index|
+      yield(self[index], index)
+    end
+  end
```
Comment (suretrust) [inline enumerables.rb]: * [x] should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_each_with_index.class != [1,2,3,4,5].each_with_index.class
```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+  def my_select
+    result = []
+    my_each do |item|
+      result << item if yield(item)
+    end
+    result
+  end
```
Comment (suretrust) [inline enumerables.rb]: * [x] should have working `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_select.class != [1,2,3,4,5].select.class
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+  def my_all?
+    result = true
+    my_each { |i| break result = false unless yield(i) }
+    result
+  end
```
Comment (suretrust) [inline enumerables.rb]: * [ ] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby 
[1, true, 'hi', []].my_all? != [1, true, 'hi', []].all?

[1,2,3,4,5].my_all?(Integer) != [1,2,3,4,5].all?(Integer)

['saheed', 'oladele', 'suretrust'].my_all?(/d/) != ['saheed', 'oladele', 'suretrust'].all?(/d/)

[1,2,3,4,5].my_all?(3) != [1,2,3,4,5].all?(3)
```

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+  def my_any?
+    result = false
+    my_each { |i| break result = true if yield(i) }
+    result
+  end
```
Comment (suretrust) [inline enumerables.rb]: * [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1, true, 'hi', []].my_any? != [1, true, 'hi', []].any?

[1,2,3,4,5].my_any?(Integer) != [1,2,3,4,5].any?(Integer)

['saheed', 'oladele', 'suretrust'].my_any?(/z/) != ['saheed', 'oladele', 'suretrust'].any?(/z/)

['saheed', 'oladele', 'suretrust'].my_any?('saheed') != ['saheed', 'oladele', 'suretrust'].any?('saheed')
```

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+  def my_none?
+    result = true
+    my_each { |i| break result = false if yield(i) }
+    result
+  end
```
Comment (suretrust) [inline enumerables.rb]: * [ ] should have working `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1, false, 'hi', []].my_none? != [1, false, 'hi', []].none?

[1,2,3,4,5].my_none?(String) != [1,2,3,4,5].none?(String)

['sure', 'trust'].my_none?(/z/) != ['sure', 'trust'].none?(/z/)

['sure', 'trust'].my_none?(5) != ['sure', 'trust'].none?(5)
```

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+  def my_count(argument)
+    counter = 0
+    my_each { |i| counter += 1 if argument == i }
+    counter
+  end
```
Comment (suretrust) [inline enumerables.rb]: * [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_count != [1,2,3,4,5].count

[1,2,3,4,5].my_count { |num| num < (1 + 5) / 2 } != [1,2,3,4,5].count { |num| num < (1 + 5) / 2 }
```

File: enumerables.rb [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    result = []
+    my_each do |item|
+      result << if block_given?
+                  yield(item)
+                else
+                  proc.call(item)
+                end
+    end
+    result
+  end
```
Comment (suretrust) [inline enumerables.rb]: * [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
[1,2,3,4,5].my_map.class != [1,2,3,4,5].map.class
```

File: enumerables.rb [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+  def my_inject(initial = nil)
+    result = initial.nil? ? self[0] : initial
+    index = initial.nil? ? 1 : 0
+    self[index...length].my_each do |item|
+      result = yield(result, item)
+    end
+    result
+  end
+end
```
Comment (suretrust) [inline enumerables.rb]: * [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
```ruby
(5..50).my_inject(4) { |prod, n| prod * n } != (5..50).inject(4) { |prod, n| prod * n }

[1,2,3,4,5].my_inject(:+) != [1,2,3,4,5].inject(:+)
```

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi @Mindo-Joseph, 👋 

### Great job thus far! 🙌 Please see my comments on your code. 👇 

- [x] There are about 20 failing cases, let's fix these.

**Kindly submit for another code review request once these have been fixed.** _Happy coding!_ 🎉

File: enumerables.rb [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+  end
+
+  def my_each_with_index
+    position = 0
+    return to_enum unless block_given?
+
+    while position <= length
+      yield self[position], position
+      position += 1
+    end
```
Comment (majovanilla) [inline enumerables.rb]: Your method is not working correctly when given a block. It should return all the elements in you enumerable and now it's not doing that.

File: enumerables.rb [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  # Define my_select
+  def my_select
+    result = []
+    return to_enum unless block_given?
+
+    my_each do |item|
+      result << item if yield item
+    end
+    result
+  end
```
Comment (majovanilla) [inline enumerables.rb:35]: Please check your method when a block is given. It's not working right now.

File: enumerables.rb [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    result = true
+    if block_given?
+      array.my_each { |x| result = false unless yield(x) }
+    elsif arg.nil?
+      array.my_each { |x| result = false unless x }
+    else
+      array.my_each { |x| result = false unless arg == x }
+    end
+    result
+  end
```
Comment (majovanilla) [inline enumerables.rb]: Please check your method when a block is given. It's not working right now. Should return true if it never returns false or nil.

Should return true when no block or argument is given and all elements are truthy.
Should return an answer if you give a class as an argument.
Should return an answer if you give a regex as an argument.
Should return an answer if you give a number as an argument.

File: enumerables.rb [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    result = false
+    if block_given?
+      array.my_each { |x| result = true if yield(x) }
+    elsif arg.nil?
+      array.my_each { |x| result = true if x }
+    else
+      array.my_each { |x| result = true if arg == x }
+    end
+    result
+  end
```
Comment (majovanilla) [inline enumerables.rb]: Please check your method when a block is given. It's not working right now. Should return true if it never returns false or nil.
Should return an answer if you give a class as an argument.
Should return an answer if you give a regex as an argument.

File: enumerables.rb [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    result = true
+    if block_given?
+      array.my_each { |x| result = false if yield(x) }
+    elsif arg.nil?
+      array.my_each { |x| result = false if x }
+    else
+      array.my_each { |x| result = false if arg == x }
+    end
+    result
+  end
```
Comment (majovanilla) [inline enumerables.rb]: Please check your method when a block is given. It's not working right now. Should return true if it never returns false or nil.
Should return an answer if you give a class as an argument.
Should return an answer if you give a regex as an argument.

File: enumerables.rb [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+    count = 0
+    if block_given?
+      my_each { |i| count += 1 if yield(i) }
+    elsif !arg.nil?
+      my_each { |i| count += 1 if arg == i }
+    else
+      return size
+    end
+    count
+  end
```
Comment (majovanilla) [inline enumerables.rb:86]: Please check your method when a block is given. It's not working right now.

File: enumerables.rb [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    array = []
+    if proc
+      my_each { |item| array << proc.call(item) }
+    elsif block_given?
+      my_each { |item| array << yield(item) }
+    else
+      return my_each
+    end
+    array
+  end
```
Comment (majovanilla) [inline enumerables.rb:98]: Should return an array with the result when you pass a block.

File: enumerables.rb [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+    result = arg2 if arg2.is_a? Numeric
+    if block_given?
+      array.my_each { |x| result = yield(result, x) }
+    elsif arg2.nil?
+      array.my_each { |x| result = result.send(arg1, x) }
+    else
+      array.my_each { |x| result = result.send(arg2, x) }
+    end
+    result
+  end
```
Comment (majovanilla) [inline enumerables.rb]: Should combine all elements when passing a block to apply an operation to them.
Should combine all elements when passing a symbol to apply an operation to them.

**Review by majovanilla — CHANGES_REQUESTED** [review]
## You did a *good* job on this project! 💜 

Some tests are still not passing. 

## Please, make all REQUIRED changes below and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

**Review by maelfosso — APPROVED** [review]
# Congratulations :clap: 

It's OK now :+1:  You can merge it now :smile: 
Good look with the next step :handshake: 

Mael FOSSO

**Pull request merged**